### PR TITLE
Migrate `random_` from the TH to Aten (CPU and CUDA)

### LINKED ISSUE #24752: Migrate `random_` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### LINKED ISSUE #32510: Tensor.random_ is not implemented for bool on CUDA(but implemented on CPU)
```
In [1]: import torch

In [2]: a = torch.empty((3, 3), dtype=torch.bool, device='cpu')

In [3]: a.random_()
Out[3]:
tensor([[False,  True, False],
        [False,  True,  True],
        [False,  True, False]])

In [4]: a = torch.empty((3, 3), dtype=torch.bool, device='cuda')

In [5]: a.random_()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-7b7864c4b98d> in <module>
----> 1 a.random_()

RuntimeError: "random_cuda_range_calc" not implemented for 'Bool'
```

cc @ngimel

### LINKED ISSUE #33299: Tensor.random_ should be able to generate all 64 bit numbers including min and max value
```
t = torch.empty(... , dtype=torch.int64)
t.random_(torch.iinfo(torch.int64).min, None)
```
should be able to fill tensor with all 64 bit number including `torch.iinfo(torch.int64).min` and `torch.iinfo(torch.int64).max`

### PR DESCRIPTION
**Pull request by pbelevich**
This pull request solves 4 problems:

1. Migrates `Tensor.random_()` from TH to ATen, both CPU and CUDA versions.
2. Allows `random_()` to generate full range of 64 bit numbers(including unsigned 64 bit max value).
3. Implements `random_()` for boolean tensors on CUDA.
4. Makes `random_()` template methods to allow using it with custom RNGs.

It is done by the following changes:

1. Drop TH CPU implementations of `random_()`.
2. Change API of `random_()` to make `to` argument optional to allow call `random_(from=min_value, to=None)` that generates full 64 bit range numbers.
3. Make three native functions `random_()`, `random_(to)` and `random_(to, from)` which call three different kernels(both CPU and CUDA).
4. Create three `random_` kernels(both CPU and CUDA) to handle `random_(no params)`, `random(to, from)` and `random(from=min_value, to=None)` cases.
5. Templatize all `random_` implementations and kernels to use them with custom RNGs.
6. Create C++ tests that uses custom RNG and `random_()` templates to check correctness.
7. Create Python tests to cover all `random_()` scenarios with all possible dtypes and devices.

Fixes #24752
Fixes #32510
Fixes #33299

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32534 Migrate `random_` from the TH to Aten (CPU and CUDA)**

Differential Revision: [D19776613](https://our.internmc.facebook.com/intern/diff/D19776613)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit e80fb25a:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e80fb25a7c7ca408e36a01808ed08e163ce5be67).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4323893) caffe2_onnx_py2_gcc5_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89859256) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89859256)) :snowflake:
<details>
<summary>
<code>Jan 23 17:07:00 unknown file: Failure</code>
</summary>

```
Jan 23 17:07:00 [       OK ] TensorIteratorTest.ComparisonLoopBinary_Short (0 ms) 
Jan 23 17:07:00 [ RUN      ] TensorIteratorTest.ComparisonLoopBinary_Int 
Jan 23 17:07:00 [       OK ] TensorIteratorTest.ComparisonLoopBinary_Int (0 ms) 
Jan 23 17:07:00 [ RUN      ] TensorIteratorTest.ComparisonLoopBinary_Long 
Jan 23 17:07:00 [       OK ] TensorIteratorTest.ComparisonLoopBinary_Long (0 ms) 
Jan 23 17:07:00 [ RUN      ] TensorIteratorTest.ComparisonLoopBinary_Float 
Jan 23 17:07:00 [       OK ] TensorIteratorTest.ComparisonLoopBinary_Float (0 ms) 
Jan 23 17:07:00 [ RUN      ] TensorIteratorTest.ComparisonLoopBinary_Double 
Jan 23 17:07:00 [       OK ] TensorIteratorTest.ComparisonLoopBinary_Double (0 ms) 
Jan 23 17:07:00 [ RUN      ] TensorIteratorTest.ComparisonLoopBinary_Bool 
Jan 23 17:07:00 unknown file: Failure 
Jan 23 17:07:00 C++ exception with description "random_ can not accept 'from' and 'to' on boolean tensors (random_ at /var/lib/jenkins/workspace/aten/src/ATen/native/Distributions.cpp:229) 
Jan 23 17:07:00 frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6a (0x7f919edda28a in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libc10.so) 
Jan 23 17:07:00 frame #1: at::native::random_(at::Tensor&, long, long, at::Generator*) + 0x1f8 (0x7f919f9233c8 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #2: <unknown function> + 0xe341a0 (0x7f919fe221a0 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #3: at::Tensor& c10::KernelFunction::callUnboxed<at::Tensor&, at::Tensor&, long, long, at::Generator*>(c10::OperatorHandle const&, at::Tensor&, long, long, at::Generator*) const + 0x5f (0x7f919f990a0f in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #4: <unknown function> + 0x27afc71 (0x7f91a179dc71 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #5: at::Tensor& c10::KernelFunction::callUnboxed<at::Tensor&, at::Tensor&, long, long, at::Generator*>(c10::OperatorHandle const&, at::Tensor&, long, long, at::Generator*) const + 0x5f (0x7f919f990a0f in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #6: at::native::randint(long, long, c10::ArrayRef<long>, at::Generator*, c10::TensorOptions const&) + 0x116 (0x7f919f986936 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #7: at::native::randint(long, long, c10::ArrayRef<long>, c10::TensorOptions const&) + 0x22 (0x7f919f986b02 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
Jan 23 17:07:00 frame #8: <unknown function> + 0xe4f953 (0x7f919fe3d953 in /var/lib/jenkins/.local/lib/python2.7/site-packages/torch/lib/libtorch_cpu.so) 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by ezyang** [comment]
Gonna let Vitaly handle this one :)

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please add benchmark numbers.

File: aten/src/ATen/native/cpu/DistributionTemplates.h [hunk 1]
```diff
@@ -8,6 +8,79 @@
...
+    });
+  } else if (isIntegralType(iter.dtype(), /*includeBool=*/true)) {
+    AT_DISPATCH_INTEGRAL_TYPES_AND(at::ScalarType::Bool, iter.dtype(), "random_cpu", [&] {
+      if (std::is_same<scalar_t, int64_t>::value) {
+        cpu_serial_kernel(iter, [generator]() -> scalar_t {
+          return generator->random64() % (static_cast<uint64_t>(std::numeric_limits<scalar_t>::max()) + 1);
+        });
+      } else if (std::is_same<scalar_t, bool>::value) {
+        cpu_serial_kernel(iter, [generator]() -> scalar_t {
+          return generator->random() & 1;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/DistributionTemplates.h]: Not sure, but can you please check distribution of booleans in randomly generated tensor

**Comment by ezyang** [comment]
Adding @anjali411 as complex may be affected by this

File: aten/src/ATen/native/Distributions.cpp [hunk 2]
```diff
@@ -213,6 +217,36 @@ Tensor& geometric_(Tensor& self, double p, Generator* gen) {
...
+Tensor& random_(Tensor& self, Generator* gen) {
+  auto iter = TensorIterator::nullary_op(self);
+  random_stub(iter.device_type(), iter, gen);
+  return self;
+}
+
+Tensor& random_(Tensor& self, int64_t from, optional<int64_t> to, Generator* gen) {
+  uint64_t range;
+  auto iter = TensorIterator::nullary_op(self);
+  if (to) {
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply aten/src/ATen/native/Distributions.cpp]: addressed

File: aten/src/ATen/native/DistributionTemplates.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+#include <ATen/native/TensorIterator.h>
+#include <c10/util/Optional.h>
+#include <limits>
+#include <cmath>
+
+namespace at {
+namespace native {
+namespace templates {
+
+template<typename scalar_t>
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/DistributionTemplates.h]: Please inline comment why we have to do it.

File: aten/src/ATen/native/cpu/DistributionTemplates.h [hunk 4]
```diff
@@ -4,9 +4,103 @@
...
+  AT_DISPATCH_ALL_TYPES(iter.dtype(), "random_full_64_range_kernel_cpu", [&] {
+    std::lock_guard<std::mutex> lock(generator->mutex_);
+    if (std::is_same<scalar_t, int64_t>::value ||
+        std::is_same<scalar_t, double>::value ||
+        std::is_same<scalar_t, float>::value) {
+      cpu_serial_kernel(iter, [generator]() -> scalar_t {
+        return static_cast<int64_t>(generator->random64()); // use all 64 bits and cast it to signed(!!!) int64_t and only after that to scalar_t implicitly
+      });
+    } else {
+      TORCH_CHECK(false, "random_full_64_range_kernel_cpu handles only int64, double and float");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/DistributionTemplates.h]: I think you need to use `TORCH_INTERNAL_ASSERT` here instead.

**Review by VitalyFedyunin — COMMENTED** [review]
Please change PR description to reflect why we are adding templates around generator.

**Comment by pbelevich** [comment]
```
import torch

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cpu')
timeit t.random_()
Before : 8.1 µs ± 29.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 7.81 µs ± 26.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cpu')
timeit t.random_()
Before : 7.2 ms ± 74.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 6.98 ms ± 17.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cpu')
timeit t.random_()
Before : 4.95 µs ± 19 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 4.85 µs ± 14.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cpu')
timeit t.random_()
Before : 4.08 ms ± 21.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 4.01 ms ± 17.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cpu')
timeit t.random_()
Before : 9.25 µs ± 223 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 8.78 µs ± 52.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cpu')
timeit t.random_()
Before : 8.08 ms ± 216 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 7.87 ms ± 13.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cpu')
timeit t.random_()
Before : 6.56 µs ± 192 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.82 µs ± 24.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cpu')
timeit t.random_()
Before : 5.5 ms ± 29 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 4.87 ms ± 18.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each

t = torch.empty(1000, dtype=torch.bool, device='cpu')
timeit t.random_()
Before : 5.44 µs ± 76.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 4.99 µs ± 21.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cpu')
timeit t.random_()
Before : 4.2 ms ± 68.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 4.1 ms ± 12.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cpu')
timeit t.random_(42)
Before : 12.1 µs ± 261 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.7 µs ± 35.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cpu')
timeit t.random_(42)
Before : 10.9 ms ± 26.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 21.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cpu')
timeit t.random_(42)
Before : 12.1 µs ± 333 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.7 µs ± 16.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cpu')
timeit t.random_(42)
Before : 10.9 ms ± 110 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 21.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cpu')
timeit t.random_(42)
Before : 12.2 µs ± 258 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.8 µs ± 35.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cpu')
timeit t.random_(42)
Before : 11 ms ± 229 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 24 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cpu')
timeit t.random_(42)
Before : 12 µs ± 24.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 12.1 µs ± 34.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cpu')
timeit t.random_(42)
Before : 11.4 ms ± 335 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 11.1 ms ± 21.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cpu')
timeit t.random_(42)
Before : 12.2 µs ± 31.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.7 µs ± 28.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cpu')
timeit t.random_(42)
Before : 11.1 ms ± 23.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 26.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cpu')
timeit t.random_(-42, 42)
Before : 11.9 µs ± 33.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.8 µs ± 36.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cpu')
timeit t.random_(-42, 42)
Before : 11 ms ± 131 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 25.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cpu')
timeit t.random_(-42, 42)
Before : 11.8 µs ± 18 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.8 µs ± 27.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cpu')
timeit t.random_(-42, 42)
Before : 10.8 ms ± 33.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 19.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cpu')
timeit t.random_(-42, 42)
Before : 12 µs ± 83 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.9 µs ± 27.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cpu')
timeit t.random_(-42, 42)
Before : 11.2 ms ± 304 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 20.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cpu')
timeit t.random_(-42, 42)
Before : 12.1 µs ± 35.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 12.2 µs ± 29 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cpu')
timeit t.random_(-42, 42)
Before : 11.1 ms ± 249 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 11.1 ms ± 34.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cpu')
timeit t.random_(-42, 42)
Before : 12.2 µs ± 35.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 11.8 µs ± 24.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cpu')
timeit t.random_(-42, 42)
Before : 11.1 ms ± 9.73 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
After  : 10.8 ms ± 49.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 8.13 µs ± 35.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 6.64 ms ± 20 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : N/A

t = torch.empty(1000000, dtype=torch.int32, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : N/A

t = torch.empty(1000, dtype=torch.float64, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 8.41 µs ± 34.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 6.92 ms ± 22.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 8.46 µs ± 40.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : 6.96 ms ± 16.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : N/A

t = torch.empty(1000000, dtype=torch.bool, device='cpu')
timeit t.random_(torch.iinfo(torch.int64).min, to=None)
Before : N/A
After  : N/A

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cuda')
timeit t.random_()
Before : 6.73 µs ± 886 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.51 µs ± 98.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cuda')
timeit t.random_()
Before : 66.4 µs ± 82.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 31.6 µs ± 41.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cuda')
timeit t.random_()
Before : 5.93 µs ± 25.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.42 µs ± 88.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cuda')
timeit t.random_()
Before : 28.2 µs ± 56.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 22.2 µs ± 121 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cuda')
timeit t.random_()
Before : 5.96 µs ± 45 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.59 µs ± 52.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cuda')
timeit t.random_()
Before : 67.3 µs ± 51.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 41 µs ± 87.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cuda')
timeit t.random_()
Before : 6.04 µs ± 18.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.48 µs ± 44.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cuda')
timeit t.random_()
Before : 28.5 µs ± 115 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 24.8 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cuda')
timeit t.random_()
Before : N/A
After  : 5.64 µs ± 27 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cuda')
timeit t.random_()
Before : N/A
After  : 21.9 µs ± 431 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cuda')
timeit t.random_(42)
Before : 8.07 µs ± 27.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.82 µs ± 73.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cuda')
timeit t.random_(42)
Before : 67.2 µs ± 97 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.5 µs ± 45.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cuda')
timeit t.random_(42)
Before : 6.07 µs ± 26.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.79 µs ± 8.44 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cuda')
timeit t.random_(42)
Before : 28.3 µs ± 42.2 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 29.9 µs ± 20.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cuda')
timeit t.random_(42)
Before : 6.1 µs ± 23.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.89 µs ± 21.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cuda')
timeit t.random_(42)
Before : 67.4 µs ± 47.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.7 µs ± 47.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cuda')
timeit t.random_(42)
Before : 6.16 µs ± 24.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.96 µs ± 21.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cuda')
timeit t.random_(42)
Before : 28.5 µs ± 70.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.3 µs ± 104 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cuda')
timeit t.random_(42)
Before : 6.14 µs ± 22.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.87 µs ± 28.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cuda')
timeit t.random_(42)
Before : 28.7 µs ± 183 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.7 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cuda')
timeit t.random_(-42, 42)
Before : 8.13 µs ± 33.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.9 µs ± 25 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cuda')
timeit t.random_(-42, 42)
Before : 67.5 µs ± 54.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 31 µs ± 59.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cuda')
timeit t.random_(-42, 42)
Before : 6.1 µs ± 16.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.98 µs ± 34.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int32, device='cuda')
timeit t.random_(-42, 42)
Before : 28.7 µs ± 108 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30 µs ± 142 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float64, device='cuda')
timeit t.random_(-42, 42)
Before : 6.26 µs ± 26.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.94 µs ± 33.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cuda')
timeit t.random_(-42, 42)
Before : 67.6 µs ± 103 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.9 µs ± 160 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cuda')
timeit t.random_(-42, 42)
Before : 8.4 µs ± 48.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 6 µs ± 28.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cuda')
timeit t.random_(-42, 42)
Before : 28.8 µs ± 148 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.3 µs ± 65.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cuda')
timeit t.random_(-42, 42)
Before : 6.15 µs ± 42.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
After  : 5.94 µs ± 18.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.bool, device='cuda')
timeit t.random_(-42, 42)
Before : 28.8 µs ± 30.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
After  : 30.7 µs ± 18.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

##############################################################

t = torch.empty(1000, dtype=torch.int64, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 6.96 µs ± 42 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.int64, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 32.7 µs ± 49.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.int32, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : N/A

t = torch.empty(1000000, dtype=torch.int32, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : N/A

t = torch.empty(1000, dtype=torch.float64, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 6.74 µs ± 30.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float64, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 32.8 µs ± 65.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.float32, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 6.66 µs ± 38.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

t = torch.empty(1000000, dtype=torch.float32, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : 32.3 µs ± 70.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

t = torch.empty(1000, dtype=torch.bool, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : N/A

t = torch.empty(1000000, dtype=torch.bool, device='cuda')
timeit t.random_(torch.iinfo(torch.int64).min, None)
Before : N/A
After  : N/A
```

**Review by VitalyFedyunin — APPROVED** [review]
LGTM with couple buts: 1) Please avoid so big PRs in the future 2) Please update docs as follow-up PR.

**Pull request closed without merging**